### PR TITLE
[Merged by Bors] - feat(Algebra/Module/CharacterModule): definition of character module

### PR DESCRIPTION
**Pull request by jjaassoonn**
For an `R`-module `M`, its character module is defined as the group homorphisms from `M` to the unit rational circle where the `R`-module structure is given by `rf(m) = f(rm)`. Let us denote the character module by `M*`

Then each linear map `f : A -> B` gives a linear map `f*` from `B*` to `A*` by composition.
When `f` is injective, `f*` is surjective.

There is a natural isomorphism between `Hom(A, B*)` and `(A ⊗ B)*`

---

- [x] depends on: #8905
- [x] depends on: #10766 (for defining the module structure)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/CharacterModule.lean [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      simp only [LinearMap.coe_comp, Function.comp_apply, dual_apply, LinearMap.id_coe, id_eq]
+      rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply]
+      erw [e.symm_apply_apply])
+
+open TensorProduct
+
+/--
+Any linear map `L : A → B⋆` induces a character in `(A ⊗ B)⋆` by `a ⊗ b ↦ L a b`
+-/
+@[simps] noncomputable def curry :
```
Comment (adamtopaz) [inline Mathlib/Algebra/Module/CharacterModule.lean:128]: I think the names `curry` and `uncurry` should be swapped. Cf. `Function.curry`.

File: Mathlib/Algebra/Module/CharacterModule.lean [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  toFun L := L.comp f.toAddMonoidHom
+  map_add' := by aesop
+  map_smul' r c := DFunLike.ext _ _ fun x ↦ by
+    simp only [RingHom.id_apply, smul_apply]
+    rw [AddMonoidHom.comp_apply, AddMonoidHom.comp_apply, smul_apply]
+    erw [smul_apply, f.map_smul]
+    rfl
+
+lemma dual_surjective_of_injective (f : A →ₗ[R] B) (hf : Function.Injective f) :
+    Function.Surjective <| dual f := by
```
Comment (alreadydone) [inline Mathlib/Algebra/Module/CharacterModule.lean:86]: This is defeq to `Module.Baer.extension_property_addMonoidHom` in my modification to #8905. For this reason I'm marking this PR dependent on #8905.

**Review by alreadydone — APPROVED** [review]
See [my branch](https://github.com/leanprover-community/mathlib4/compare/zjj/CharacterModuleDefsAndBasicAPI...jyxu/CharacterModule) for golfs.

File: Mathlib/Algebra/Module/CharacterModule.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/Algebra/Module/CharacterModule.lean]: thanks

File: Mathlib/Algebra/Module/CharacterModule.lean (hunk 1 above)
Comment (jjaassoonn) [reply Mathlib/Algebra/Module/CharacterModule.lean]: done

File: Mathlib/Algebra/Module/CharacterModule.lean (hunk 2 above)
Comment (alreadydone) [reply Mathlib/Algebra/Module/CharacterModule.lean]: Will you incorporate [these golfs](https://github.com/leanprover-community/mathlib4/compare/zjj/CharacterModuleDefsAndBasicAPI...jyxu/CharacterModule) too?

Comment (jjaassoonn) [reply Mathlib/Algebra/Module/CharacterModule.lean]: Thanks for the golfing. The `$()` trick is amazing, what do they do please.

Comment (alreadydone) [reply Mathlib/Algebra/Module/CharacterModule.lean]: The congr elaborator was added in [#2544](https://github.com/leanprover-community/mathlib4/commit/9a9e31c76cebda72027752554b7da6c991fcc231) and allows you to combine multiple congr_arg and (DFunLike.)congr_fun into a single `congr()` expression; the equality (or equalities) involved need to be labeled with `$`.

**Review by alreadydone — APPROVED** [review]
Thanks!
maintainer merge

**Comment by adamtopaz** [comment]
Thanks!
bors r+

**Pull request closed without merging**